wagtail/models/__init__.py

```diff
@@ -1299,6 +1299,143 @@ def __init__(self, *args, **kwargs):
     def __str__(self):
         return self.title
 
+    def _process_leaf(self, nodes):
+        """
+        Process children to add to a leaf node
+        """
+        # Get number of children
+
+        step = 1
+        max_length = self.__class__._meta.get_field("path").max_length
+        for node in nodes:
+            node.depth = self.depth + 1
+            node.path = self.__class__._get_path(self.path, node.depth, step)
+            step += 1
+            if len(node.path) > max_length:
+                raise ValidationError(
+                    "The new node is too deep in the tree, try \
+                      increasing the path.max_length property \
+                    and UPDATE your database"
+                )
+            self.numchild += 1
```

Comment:
It would be a nice touch if in-memory changes were also reverted in the event of an error.